File: app/src/main/java/com/popalay/cardme/business/exception/ExceptionFactory.java
package com.popalay.cardme.business.exception;

import android.support.annotation.StringRes;

import com.popalay.cardme.R;

public final class ExceptionFactory {

    private ExceptionFactory() {
    }

    public enum ErrorType {
        CARD_EXIST,
        PERMISSION_DENIED
    }

    public static Throwable createError(ErrorType type) {
        return createError(type, R.string.none);
    }

    public static Throwable createError(ErrorType type, @StringRes int message) {
        return new AppException(type, message);
    }

    public static boolean isAppException(Throwable throwable) {
        return throwable instanceof AppException;
    }
}

File: app/src/main/java/com/popalay/cardme/business/exception/ExceptionFactory.kt
package com.popalay.cardme.business.exception

import android.support.annotation.StringRes

import com.popalay.cardme.R

object ExceptionFactory {

    enum class ErrorType {
        CARD_EXIST,
        PERMISSION_DENIED
    }

    fun createError(type: ErrorType, @StringRes message: Int = R.string.none) = AppException(type, message)

    fun isAppException(throwable: Throwable) = throwable is AppException
}
